### PR TITLE
Scheduled tasks: add feature to scan for network devices before shutting down

### PR DESCRIPTION
**Pull request by p-betula-pendula**
[Forum entry](https://forum.rockstor.com/t/suspend-shutdown-when-certain-devices-are-offline-using-scheduled-tasks-2038) and issue #2038.

This needs to be looked through because of my lack of experience.

especially:
- check grammar and phrasing in the UI (and code)
- linting
- best practices I'm not aware of
- possible conflicts with other components

It's working flawlessly when tested in my virtual environment.

OpenSUSE is yet to be tested.

### CODE REVIEW

File: src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py [hunk 1]
```diff
@@ -33,6 +38,45 @@ def validate_shutdown_meta(meta):
...
+                if not all_devices_offline(addresses):
+                    logger.debug('At least one pinged device is still online')
+                    return False
+                else:
+                    # don't sleep after last scan
+                    if (i+1) >= iterations:
+                        break
+                    logger.debug('No device answered, trying again in ' + str(interval) + ' seconds')
+                    # simply sleep as done in pool_scrub.py
+                    time.sleep(interval)
```
Comment (p-betula-pendula) [inline src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py]: Is using time.sleep() okay in this context?

**Comment by phillxnet** [comment]
@p-betula-pendula Well done on getting to the pull request stage with this one.

We do need this tested in openSUSE Tumbleweed first (and ideally also Leap 15.1) but yes it shouldn't make any difference but just in case. We have had some unexpected surprises so best we stick to that during this transition phase. I'll mark as need review.

**Comment by p-betula-pendula** [comment]
I've successfully tested both cases (shutting down and staying online) in Leap 15 using a clean install with Rockstor built from source. The ping command behaves as expected and I bet it does so for Tumbleweed as well – unfortunately, I've picked the less important of the two as it seems. I'll try to test on Tumbleweed soon.

**Comment by p-betula-pendula** [comment]
Successfully tested on openSUSE Tumbleweed. The code has now been tested on all three distributions.

**Comment by phillxnet** [comment]
@p-betula-pendula This issue is still awaiting review from a prior rockstor-core contributor so hang in there and we can hopefully get it in there after. Although as we have quite a few outstanding pull requests it may take a release or two before we get to it. Sorry for the delay.
Calling on prior rockstor-core contributors to review this pull request. I can't check it myself just yet but will do when next my backlog allows. We may also have a pending merge difficult on the horizon as there is also a major pull request that should precede this one as it represents a bug fix. But we can cross that one when we come to it.

Calling for a review here as this is a nice feature and it would be nice to have it reviewed ready for merge. Please note @p-betula-pendula query re the contextual use of time.sleep().

File: src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py [hunk 2]
```diff
@@ -33,6 +38,45 @@ def validate_shutdown_meta(meta):
     return meta
 
 
+def all_devices_offline(addresses):
+    for address in addresses:
+        logger.debug('Pinging ' + address + '...')
```
Comment (FroggyFlox) [inline src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py]: Simple style reformatting:
```suggestion
        logger.debug('Pinging {}...'.format(address))
```

File: src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py [hunk 3]
```diff
@@ -33,6 +38,45 @@ def validate_shutdown_meta(meta):
...
+
+        if len(addresses) > 0:
+            addresses = addresses[0]
+            # remove all non word characters except from '.' and '-' to prevent
+            # the execution of shell code when running the ping command
+            addresses = [re.sub(r"[^\w\.-]", '', address) for address in addresses]
+            interval = float(meta['ping_scan_interval']) if float(meta['ping_scan_interval']) >= 5 else 5.0
+            iterations = int(meta['ping_scan_iterations']) if int(meta['ping_scan_iterations']) >= 1 else 1
+
+            logger.debug('Pinging devices: '+str(addresses))
```
Comment (FroggyFlox) [inline src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py]: Small style reformatting:
```suggestion
            logger.debug('Pinging devices: {}'.format(addresses))
```

File: src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py [hunk 4]
```diff
@@ -33,6 +38,45 @@ def validate_shutdown_meta(meta):
...
+            for i in range(iterations):
+                if not all_devices_offline(addresses):
+                    logger.debug('At least one pinged device is still online')
+                    return False
+                else:
+                    # don't sleep after last scan
+                    if (i+1) >= iterations:
+                        break
+                    logger.debug('No device answered, trying again in ' + str(interval) + ' seconds')
```
Comment (FroggyFlox) [inline src/rockstor/scripts/scheduled_tasks/reboot_shutdown.py]: Small style reformatting:
```suggestion
                    logger.debug('No device answered, trying again in {} seconds.'.format(interval))
```

File: src/rockstor/system/osi.py [hunk 5]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
 
+def is_network_device_responding(address):
+    """
+    Small function that sends ICMP echo requests to a given network device,
+    either using the hostname or IP address.
+    :param address: the hostname or IP address of the device to be pinged
+    :return: true if the device is responding, false if not
+    """
+    # because of -c 3, three requests will be sent
+    output, error, return_code = run_command([PING, '-c 3 -q', address], log=False, throw=False)
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Style reformatting + fix for unrecognized command in  up-to-date Tumbleweed:
```suggestion
    cmd = [PING, '-c', 3, '-q', address]
    o, e, rc = run_command(cmd, log=True, throw=False)
```

File: src/rockstor/system/osi.py [hunk 6]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+    # The ping command will always return 0 if at least one request has been answered,
+    # 1, if no request has been answered and another error code if the hostname is not known.
+    # The corresponding error message is 'ping: <device name>: Name or service not known'.
+    if return_code == 0:
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
    if rc == 0:
```

File: src/rockstor/system/osi.py [hunk 7]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+        return True
+    elif return_code == 1 or \
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
    elif rc == 1 or \
```

File: src/rockstor/system/osi.py [hunk 8]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+            next((s for s in output if '0 received' in s), None) or \
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
            next((s for s in o if '0 received' in s), None) or \
```

File: src/rockstor/system/osi.py [hunk 9]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+            next((s for s in error if 'ping: ' + str(address) in s), None):
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
            next((s for s in e if 'ping: {}'.format(address) in s), None):
```

File: src/rockstor/system/osi.py [hunk 10]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+        return False
+    logger.debug('Ping command unexpectedly exited with return code ' + str(return_code))
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
    logger.debug('Ping command unexpectedly exited with return code {}'.format(return_code))
```

File: src/rockstor/system/osi.py [hunk 11]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+    if len(error):
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
    if len(e):
```

File: src/rockstor/system/osi.py [hunk 12]
```diff
@@ -1203,6 +1203,30 @@ def set_system_rtc_wake(wakeup_epoch):
...
+        logger.debug(error[0])
```
Comment (FroggyFlox) [inline src/rockstor/system/osi.py]: Small style reformatting
```suggestion
        logger.debug(e[0])
```

**Comment by p-betula-pendula** [comment]
> Hi @p-betula-pendula and thank you for this feature, that looks really nice!
> 
> I've tested it in the latest Tumbleweed (20190708) but although no errors were thrown, I noticed my test network device was not detected as online despite _actually_ being online (as verified by manual `ping` command). After a quick verification (switching to `Log=True` in `run_command`), it turns out the `-c 3 -q` string was not recognized as a valid argument for the `ping` command.
> I thus reformatted it to be more harmonized with the rest of Rockstor and it now works as expected.

Thanks for testing it out and eliminating the problem!

**Comment by phillxnet** [comment]
@p-betula-pendula Just a quick update.
@FroggyFlox has done a review and we are now just waiting for you to commit their suggestions, which is most easily done via GitHubs "Commit suggestions" buttons. Would be great to have this one ready to go in so if you are game to commit those changes we should have nudged this one much nearer the edge. Passing up in this next merge round but at least their after we should have some more network unit testing in place (from @FroggyFlox) that can only help with proving this functionality stay functional into the future. We also have:

> @phillxnet, should we include a (series of) unit test(s) for the new function `is_network_device_responding()` in `system/osi.py`? So far, all pre-existing `test_osi*` pass.

Not sure if this will need a rebase by the time the next merge round comes but we can cross that bridge when we get to it.

So in the mean time if you could commit @FroggyFlox's suggestions and we can then take it from there. Thanks a tone for this rather snazy feature and I'm keen to get it in. But unfortunately our recent server migration has put me rather behind schedule so I'm playing catch up currently.

**Comment by p-betula-pendula** [comment]
@phillxnet My apologies, I forgot I need to commit the suggestions for you to be able to merge them. I've committed them just now.

I'm looking forward to this feature being available in one of the next releases. Thanks for your commitment to Rockstor!

**Comment by phillxnet** [comment]
@p-betula-pendula This is a really neat addition. Thanks, and nicely implemented.

@FroggyFlox Thanks for giving this a thorough look over and offering up the tweaks/fixes etc. Much appreciated and a great help.

Again apologies for the delay getting this one in. I'm initially anticipating it's appearance in our stable channel updates version 3.9.2-51; if all goes to plan. This should be released shortly once I've looked over and hopefully merged another pr or two.

Thanks for your patience and for persevering with this pull request, and for the debug messages also: nice.

**Comment by phillxnet** [comment]
@p-betula-pendula Re:

> I'm initially anticipating it's appearance in our stable channel updates version 3.9.2-51; if all goes to plan.

Apart from a much larger delay than I had anticipated, 3.9.2-51 has now been released, with your much appreciated enhancement in place. Thanks again for your efforts and patience. The release announcement has now been posted on the forum:
https://forum.rockstor.com/t/3-9-2-stable-channel-changelog/5741/18
and each updating system should display a list of fixes / enhancements along with their respective contributor links via the Web-UI.
Also keep an eye on the forum for a pending testing channel announcement.

**Pull request merged**